Here I use tensorflow to write functions, you can set any number of hiddens layers and nodes. Also, you can set your learning rate. So it is easier for you to try different network structures. 

I only run my main functions. I change Mnist data for train and test. You can change your own data for classification and have a try.

In [1]:
import pandas as pd
import numpy as np
from pandas import DataFrame
from functions import array_to_one_hot
from session_run import session_run

train = pd.read_csv("mnist_train.csv")
test = pd.read_csv("mnist_test.csv")

# see training and test data

In [3]:
train.head()
# I change label 0-9 to 10-19 for you to distinguish predict index and label

class  x_0  x_1  x_2  x_3  x_4  x_5  x_6  x_7  x_8  ...    x_774  x_775  \
0     10  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...      0.0    0.0   
1     10  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...      0.0    0.0   
2     10  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...      0.0    0.0   
3     10  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...      0.0    0.0   
4     10  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...      0.0    0.0   

   x_776  x_777  x_778  x_779  x_780  x_781  x_782  x_783  
0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0  
1    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0  
2    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0  
3    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0  
4    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0  

[5 rows x 785 columns]

I use 1000 traing and 200 test for each class

In [6]:
train['class'].value_counts()

15    1000
14    1000
13    1000
12    1000
19    1000
11    1000
18    1000
10    1000
17    1000
16    1000
Name: class, dtype: int64

In [8]:
test['class'].value_counts()

19    200
18    200
17    200
16    200
15    200
14    200
13    200
12    200
11    200
10    200
Name: class, dtype: int64

# prepare data for tensorflow input form

In [10]:
## we random shuffle the train data
train = train.sample(frac=1).reset_index(drop=True)

feature_name = train.columns[1:]
# always set class label in the first column
label_name = train.columns[0] 

train_feature = train[feature_name].as_matrix()
test_feature = test[feature_name].as_matrix()

labels = sorted(set(train[label_name].tolist()))

# remember to change label to one-hot form, this is a very important step
train_label_one_hot = array_to_one_hot(train[label_name].tolist(),labels)
test_label_one_hot = array_to_one_hot(test[label_name].tolist(),labels)

train_label_array = np.array(train[label_name])
test_label_array = np.array(test[label_name])

In [13]:
train_feature

array([[ 0.,  0.,  0., ...,  0.,  0.,  0.],
       [ 0.,  0.,  0., ...,  0.,  0.,  0.],
       [ 0.,  0.,  0., ...,  0.,  0.,  0.],
       ..., 
       [ 0.,  0.,  0., ...,  0.,  0.,  0.],
       [ 0.,  0.,  0., ...,  0.,  0.,  0.],
       [ 0.,  0.,  0., ...,  0.,  0.,  0.]])

In [16]:
train_label_one_hot

array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 1, 0],
       [1, 0, 0, ..., 0, 0, 0],
       ..., 
       [0, 0, 0, ..., 0, 1, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]])

In [18]:
print("feature input shape:",train_feature.shape)
print("class input shape:",train_label_one_hot.shape)

feature input shape: (10000, 784)
class input shape: (10000, 10)


# set tensorflow parameters

Learning rate is 0.01 and I train for 10 times and display result for each step.

In [19]:
# Parameters
train_num = train_feature.shape[0] ## train_size
learning_rate = 0.01
training_epochs = 10
batch_size = 100
display_step = 1

You can set any number of hidden layers and nodes. Here I use two hidden layers, first layer have 1600 nodes and second layer have 800 nodes.

In [21]:
# Network Parameters
layers_structure = [1600,800]

In [22]:
run = session_run(train_feature, train_label_one_hot, train_label_array, \
                test_feature, test_label_one_hot, test_label_array, labels,\
                layers_structure, train_num, learning_rate, training_epochs,\
                batch_size, display_step)

period 0 train: 0.9462 test: 0.9365
period 1 train: 0.967 test: 0.965
period 2 train: 0.9695 test: 0.9665
period 3 train: 0.9809 test: 0.9775
period 4 train: 0.9852 test: 0.979
period 5 train: 0.9845 test: 0.978
period 6 train: 0.9801 test: 0.9815
period 7 train: 0.9907 test: 0.992
period 8 train: 0.9873 test: 0.9875
period 9 train: 0.991 test: 0.994


You can see after 10 times training, accuarcy is about 99.4%.
Below is the session_run function return. You can see training and test accuracy, predict label, predict probability matrix.
```python
  return_dict = {"test_accuracy":test_accuracy,
                   "train_accuracy":train_accuracy,
                   "train_pred_label_array":train_pred_label_array,
                   "test_pred_label_array":test_pred_label_array,
                   "train_pred_prob_matrix":train_pred_prob_matrix,
                   "test_pred_prob_matrix":test_pred_prob_matrix,
                   "test_loss":test_loss,
                   "train_loss":train_loss}
```            

In [31]:
print("test_accuracy:",run['test_accuracy'])
print("train_accuracy:",run['train_accuracy'])
print("test_pred_prob_matrix",run['test_pred_prob_matrix'])

test_accuracy: 0.994
train_accuracy: 0.991
test_pred_prob_matrix [[  1.00000000e+00   0.00000000e+00   0.00000000e+00 ...,   0.00000000e+00
    0.00000000e+00   0.00000000e+00]
 [  1.00000000e+00   0.00000000e+00   0.00000000e+00 ...,   0.00000000e+00
    0.00000000e+00   0.00000000e+00]
 [  1.00000000e+00   0.00000000e+00   0.00000000e+00 ...,   0.00000000e+00
    0.00000000e+00   0.00000000e+00]
 ..., 
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00 ...,   2.42040168e-16
    0.00000000e+00   1.00000000e+00]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00 ...,   0.00000000e+00
    0.00000000e+00   1.00000000e+00]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00 ...,   0.00000000e+00
    0.00000000e+00   1.00000000e+00]]


In [32]:
print("train_pred_label_array:",run['train_pred_label_array'])
print("test_pred_label_array:",run['test_pred_label_array'])

train_pred_label_array: [13 18 10 ..., 18 15 14]
test_pred_label_array: [10 10 10 ..., 19 19 19]
